src/ray/raylet/node_manager.cc

```diff
@@ -312,61 +458,133 @@ void NodeManager::SubmitTask(const Task &task, const Lineage &uncommitted_lineag
   }
 }
 
-void NodeManager::AssignTask(const Task &task) {
+void NodeManager::AssignTask(Task &task) {
+  const TaskSpecification &spec = task.GetTaskSpecification();
+
+  // If this is an actor task, check that the new task has the correct counter.
+  if (spec.IsActorTask()) {
+    if (CheckDuplicateActorTask(actor_registry_, spec)) {
+      // Drop tasks that have already been executed.
+      return;
+    }
+  }
+
   // Resource accounting: acquire resources for the scheduled task.
   const ClientID &my_client_id = gcs_client_->client_table().GetLocalClientId();
-  RAY_CHECK(this->cluster_resource_map_[my_client_id].Acquire(
-      task.GetTaskSpecification().GetRequiredResources()));
-
-  if (worker_pool_.PoolSize() == 0) {
-    worker_pool_.StartWorker();
+  RAY_CHECK(
+      this->cluster_resource_map_[my_client_id].Acquire(spec.GetRequiredResources()));
+
+  // Try to get an idle worker that can execute this task.
+  std::shared_ptr<Worker> worker = worker_pool_.PopWorker(spec.ActorId());
+  if (worker == nullptr) {
```

Comment:
Is this a thing that could happen a lot? E.g., would `AssignTask` be called if there are no workers available?

Reply:
It can happen pretty often, either at the beginning when workers might not have started/connected yet, or during execution when existing workers are blocked in a `ray.get`.